[Snackbar] Allow Snackbars to have leading alignment on iPads

andrewoverton opened the pull request:
Snackbars are allowed to have leading alignment on iPads, but until now no implementations have been suggested. Here's an idea for how it could be done using an alignment enum.

If merged this will close issue #2830 and resolve pivotal ticket [156805088](https://www.pivotaltracker.com/n/projects/2150420/stories/156805088).

Here's a screenshot:
![simulator screen shot - ipad air 2 - 2018-05-14 at 13 26 01](https://user-images.githubusercontent.com/8020010/40019940-b510d4e0-578e-11e8-8b6f-4adda8ffadb3.png)

andrewoverton committed:
Add leadingAlignment for snackbars on iPad

yarneo commented:
Congrats on your first PR!! 🎉🎉🎉🎉

andrewoverton committed:
Make center and leading constraints nullable

andrewoverton committed:
Add/remove necessary/unnecessary import statements

jverkoey commented on components/Snackbar/src/MDCSnackbarAlignment.h:
```diff
@@ -0,0 +1,30 @@
+  // and only permitted setting for iPhone.
+  MDCSnackbarAlignmentCenter = 0,
+
+  // Snackbar is positioned near the leading margin.
```
Please the comments change here and above to `/** */`.

Context: public APIs require use of `/** */` tags in order to have the docs picked up by our API doc generator.

jverkoey commented on components/Snackbar/src/MDCSnackbarAlignment.h:
```diff
@@ -0,0 +1,30 @@
+/**
+ States used to configure snackbar alignment.
+ */
+typedef NS_ENUM(NSInteger, MDCSnackbarAlignment) {
```
I can't tell at first glance whether this is alignment within the snackbar (e.g. UITextAlignment aligns the text within the label's bounds) or alignment to the screen. `MDCSnackbarScreenAlignment` might help clarify the intended use of this API. What do you think?

andrewoverton replied:
I was using `NSTextAlignment` as a guide. `NSTextAlignment` aligns _text_ in a container, like a `UILabel`, for example. Similarly, I figured `MDCSnackbarAlignment` could align a snackbar in a container. I'm open to `MDCSnackbarScreenAlignment` though, definitely.

jverkoey replied:
Ah, interesting. I just realized that when I look at MDC APIs I tend to subconsciously drop both the MDC and the component name, so I read this API as simply “alignment” and wasn’t sure what was being aligned.

NSTextAlignment on the other hand I read as “text alignment”.

This may just be me though, so I don’t feel super strongly about adding “screen” to the API.

andrewoverton replied:
It wasn't brought up in any of the previous comments--how do we feel about introducing such a big API addition (a whole new header file!) for something that was originally labeled "good first issue"? It seemed a little bold to me, I just couldn't think of a better way haha 🙈

romoore replied:
I wouldn't worry about a new header too much. Since it's used by both SnackbarManager and SnackbarMessageView (and neither publicly references the other), I think it's good to put it in a separate header.

@jverkoey If we want to later change the alignment of the content within the Snackbar, I think we could use something like `MDCSnackbarContentAlignment` to name that value.

jverkoey commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -77,6 +78,17 @@
 
+/**
+ Determines whether the snackbar is presented in the center of the of the screen or (on iPad only)
+ from the leading edge of the screen.
```
I fear this documentation will get out of date fairly quickly — can we future-proof it by removing references to the specific values it supports?

E.g. "Determines the snackbar alignment to the screen."

andrewoverton replied:
Oooh, good call 👍

jverkoey commented on components/Snackbar/src/MDCSnackbarAlignment.h:
```diff
@@ -0,0 +1,30 @@
+typedef NS_ENUM(NSInteger, MDCSnackbarAlignment) {
+
+  // Snackbar is positioned in the center of the screen. This is the default setting
+  // and only permitted setting for iPhone.
```
I can imagine us wanting to support alternative alignments on large landscape phones in the future, so it may be wise for us to avoid restricting when/how to use the alignments. What do you think?

andrewoverton replied:
Are you suggesting our implementation allow behavior that goes against the spec? I'm fine with that--I honestly wasn't sure how much I should enforce adherence.. Would you recommend I go so far as to include a _trailing_ alignment case, even though it's a definite no-no, from what I understand? I can see why it might be a wise move for future-proofing.

jverkoey replied:
Generally-speaking we’ve been moving toward increased configurability and decreased opinions baked in to the components themselves.
This is motivated by the fact that configuration is almost always a need, while opinions tend to change from one design system to another. This trend is best exemplified in the themer APIs, which apply a set of opinions on configurable components.

As for adding additional behavior - I think it’s best to keep the scope of this change as-is for now. If someone wants a trailing implementation in the future then we can/will add it then.

jverkoey replied:
Thinking more about this though...I wonder if we do want to encourage thoughtful use of this API on a traits basis? Perhaps by somehow specifying which traits to use a given alignment on?

andrewoverton replied:
Sounds good! Thanks for the insight :D

andrewoverton replied:
Oh, I just noticed https://github.com/material-components/material-components-ios/pull/4163#discussion_r188414240. Hmmmm.

jverkoey commented on components/Snackbar/src/private/MDCSnackbarOverlayView.h:
```diff
@@ -62,4 +63,15 @@ OBJC_EXTERN NSTimeInterval const MDCSnackbarLegacyTransitionDuration;
+ If called within an animation block, the change will be animated.
+
+ @note This method must be called from the main thread.
+ @warning On non-iPad devices this setter will be a no-op.
```
Drop this comment - we generally always want properties to store the provided state even if they're not making use of it.

jverkoey commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -246,7 +246,7 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
     CGFloat bottomMargin = [self staticBottomMargin];
     CGFloat sideMargin = [self sideMargin];
-    BOOL fullWidth = UI_USER_INTERFACE_IDIOM() != UIUserInterfaceIdiomPad;
+    BOOL isIphone = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone;
```
Should we be looking at the size traits of the device? E.g. a large iPhone in landscape is almost comparable to an iPad in portrait.

romoore replied:
Agreed, please consider using horizontal Regular traits instead of iPad. An iPad in 1/3 split-screen multitasking is basically a tall iPhone.

romoore replied:
Or even Regular/Regular if you want to prohibit it in a vertically-compact environment (like a landscape iPhone+).

https://useyourloaf.com/blog/size-classes/

andrewoverton replied:
I agree that size classes are the way to go--I'll look into it!

yarneo replied:
We don't want to break appearance by changing the width from full to non-full on mobile devices for landscape. Currently the behavior and per spec is to have it full width, i think we should leave it at that rather than making breaking layout changes.

jverkoey commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -617,6 +627,33 @@ - (void)setBottomOffset:(CGFloat)bottomOffset {
+- (void)setAlignment:(MDCSnackbarAlignment)alignment {
+  if (UI_USER_INTERFACE_IDIOM() != UIUserInterfaceIdiomPad) {
+    return;
+  }
```
Drop this early exit - we always want to set the _alignment ivar so that the state isn't lost.

andrewoverton replied:
I think this makes sense if we actually are okay with people doing things like creating leading aligned snackbars in horizontally compact environments. It sounds like we are 👍

romoore replied:
I think the guidelines are pretty clear that it's only for "wide layouts". I would not consider a Compact size class to be wide.  An iPhone 4 is probably far too narrow to consider anything other than a full-screen Snackbar.

andrewoverton replied:
So when someone tries to set a leading alignment in a horizontally compact environment we want to preserve the state (we want the getter to return `MDCSnackbarAlignmentLeading`) but we want the snackbar to _appear_ centered/full-screen. Okie dokie, thanks for the clarification  👍

jverkoey reviewed (dismissed):
Love the constraint-based solution for layout. I have some small questions about the implementation but overall this is looking like a good direction.

andrewoverton committed:
Address PR feedback

andrewoverton commented:
Here's a gif of the size class based solution added in 8ce8cfe. In this case the alignment is leading for both snackbars shown.
![may-15-2018 17-33-57](https://user-images.githubusercontent.com/8020010/40119699-56a91bb6-58eb-11e8-9cc4-e42089c1dead.gif)

yarneo reviewed (commented):
can you please provide screenshots and test out the new material snackbar. You can look at it by pressing the top right nav button in the example you are currently screenshotting. Also please try it out with dynamic type on to see it is working well.

yarneo commented:
Waiting for design feedback on if landscape on big iPhones is considered a "wide UI". Will follow up.

andrewoverton commented:
@yarneo thanks for asking an actual designer for clarification, I should've thought to do that 😛 

Here is a gif with dynamic type and the new snackbar:

![may-16-2018 10-46-46](https://user-images.githubusercontent.com/8020010/40124423-8bfa5bc6-58f6-11e8-8603-4c6c4d1f3c93.gif)

andrewoverton committed:
Only allow leading aligned snackbar with reg h & w

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+                                                                          toItem:container
+                                                                       attribute:NSLayoutAttributeLeading
+                                                                      multiplier:1.0
+                                                                        constant:sideMargin];
```
I like the use of Regular/Regular here, and I think it aligns with the Material Guidelines. Looks like sideMargin is dependent on device idiom (iPad/iPhone). Should we migrate that code to size classes as well? I'm concerned that a split-screen iPad in a Regular(H)/Compact(W) environment will have the wrong margins (24 instead of 8).

andrewoverton replied:
I can do that!

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+
+        // If not full width, ensure that it doesn't get any larger than our own width.
+        [container
+         addConstraint:[NSLayoutConstraint constraintWithItem:snackbarView
```
Nit/Indentation: Should be 4 spaces from the previous line.

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+
+        // Also ensure that it doesn't get any smaller than its own minimum width.
+        [container
+         addConstraint:[NSLayoutConstraint constraintWithItem:snackbarView
```
Nit/Indentation: Same (4 spaces)

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+
+        // Also ensure that it doesn't get any larger than its own maximum width.
+        [container
+         addConstraint:[NSLayoutConstraint constraintWithItem:snackbarView
```
Nit/Indentation: Same as above.

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
-            rightMargin += self.safeAreaInsets.right;
-          }
+        if (@available(iOS 11.0, *)) {
+          leftMargin += self.safeAreaInsets.left;
```
Nit/Suggestion: Consider adding an mdc_safeAreaInsets method similar to [MDCBottomAppBarView](https://github.com/material-components/material-components-ios/blob/develop/components/BottomAppBar/src/MDCBottomAppBarView.m#L251) and [MDCBottomNavigationBar](https://github.com/material-components/material-components-ios/blob/develop/components/BottomNavigation/src/MDCBottomNavigationBar.m#L355).  It will simplify this code (slightly) and allow us to easily migrate in the future.

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -257,85 +257,79 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+                                                                 toItem:container
+                                                              attribute:NSLayoutAttributeLeading
+                                                             multiplier:1.0
+                                                               constant:leftMargin]];
```
RTL: Please be sure to flip `leftMargin` and `rightMargin` in RTL layouts.  I believe `safeAreaInsets` should always be set correctly for left/right, but the "leading" constraint in an RTL layout should use the `rightMargin` value (since that edge is on the right), not the `leftMargin` value.

romoore commented on components/Snackbar/src/MDCSnackbarAlignment.h:
```diff
@@ -0,0 +1,33 @@
+#import <Foundation/Foundation.h>
+
+/**
+ States used to configure snackbar alignment.
```
Super-nit: s/snackbar/Snackbar/

romoore commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -77,6 +78,15 @@
+
+ @note This method must be called from the main thread.
+ */
++ (void)setAlignment:(MDCSnackbarAlignment)alignment;
```
Please consider making this a property so clients can read the value as well.

andrewoverton replied:
Since it's a class method that sets a property on the internal snackbar manager would you mind just another method that reads the value from the internal snackbar manager?

romoore replied:
Maybe a class property would work well?

andrewoverton replied:
Ah, I forgot that class properties were added to Objective-C in 2016, nice

romoore commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -77,6 +78,15 @@
+
+ If called within an animation block, the change will be animated.
+
+ @note This method must be called from the main thread.
```
Nit: You can drop the main thread `@note` since this is a UIView, clients are expected to do all UI-related work on the main thread.

andrewoverton replied:
Ah, ok, I was just going off what I thought were conventions based off my observations.

jverkoey replied:
For things where there are existing conventions we do prefer staying consistent rather than fixing in-place so as not to increase the scope of PRs. @romoore / @andrewoverton this looks like a good thing to file an issue for to do follow-up cleanup work.

andrewoverton replied:
Alrighty then I will add them back!

romoore commented on components/Snackbar/src/MDCSnackbarAlignment.h:
```diff
@@ -0,0 +1,33 @@
+typedef NS_ENUM(NSInteger, MDCSnackbarAlignment) {
+  /**
+   Snackbar is positioned in the center of the screen. This is the default setting
+   and only permitted setting for iPhone.
```
Nit: Consider putting the default/note about size classes in the docs for the property itself rather than on the enumeration.

```objectivec
/**
 Determines the Snackbar alignment to the screen. If called within an animation block, the change 
 will be animated.

 @note This setting is only used when both the horizontal and vertical size classes of the presenting
       window are @c UIUserInterfaceSizeClassRegular. Otherwise @c MDCSnackbarAlignmentCenter 
       will be used.
 */
```

romoore commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -534,6 +534,13 @@ + (void)setBottomOffset:(CGFloat)offset {
 }
 
++ (void)setAlignment:(MDCSnackbarAlignment)alignment {
+  NSAssert([NSThread isMainThread], @"setAlignment must be called on main thread.");
```
Nit: I think this can be removed, clients are expected to perform UI work on the main thread.

romoore commented on components/Snackbar/src/MDCSnackbarMessageView.m:
```diff
@@ -924,6 +924,22 @@ - (NSArray *)horizontalButtonLayoutConstraints {
+      break;
+    default:
+      NSAssert(NO, @"`alignment` should be a valid MDCSnackbarAlignment value");
+      break;
```
Please add default behavior here for non-debug releases. I would propose the `MDCSnackbarAlignmentCenter` behavior should be used.

romoore commented on components/Snackbar/src/MDCSnackbarMessageView.m:
```diff
@@ -956,7 +972,7 @@ - (CGFloat)contentSafeBottomInset {
   // If a bottom offset has been set to raise the HUD, e.g. above a tab bar, we should ignore
```
Super-nit: s/HUD/Snackbar/

romoore commented on components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h:
```diff
@@ -53,7 +53,19 @@ typedef void (^MDCSnackbarMessageDismissHandler)(BOOL userInitiated,
+@property(nonatomic) BOOL anchoredToScreenBottom;
+
+/**
+ The layout constraint which determines how far the snackbar is from the leading edge of the screen.
```
Nit: %s/snackbar/Snackbar/g

(apply change throughout)

andrewoverton replied:
~I wouldn't want the diff to get significantly larger so how about I do this just in my changes rather than for all other lowercase mentions of Snackbars?~

romoore replied:
Agreed, sorry for the confusion (I wrote it poorly).

andrewoverton replied:
I already did it for a lot of occurrences... I can push it up and you can let me know if you want me to re-lowercase them...

romoore replied:
Oh, it's fine. I meant that you didn't need to do unnecessary work. 🤷‍♀️

romoore commented on components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h:
```diff
@@ -53,7 +53,19 @@ typedef void (^MDCSnackbarMessageDismissHandler)(BOOL userInitiated,
+
+/**
+ The layout constraint which determines how far the snackbar is from the leading edge of the screen.
+ It is active when the alignment of the parent overlay view is MDCAlignmentLeading.
```
s/MDCAlignmentLeading/MDCSnackbarAlignmentLeading/

romoore commented on components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h:
```diff
@@ -53,7 +53,19 @@ typedef void (^MDCSnackbarMessageDismissHandler)(BOOL userInitiated,
+
+/**
+ The layout constraint used to center the snackbar.
+ It is active when the alignment of the parent overlay view is MDCAlignmentCenter.
```
s/MDCAlignmentCenter/MDCSnackbarAlignmentCenter/

andrewoverton replied:
Good catch!

romoore reviewed (changes requested):
Overall looks good to me. A few comments and a handful of nits.  Please also add tests (unit or UI) or file a backlog issue so we can track adding tests.

andrewoverton committed:
Address more PR feedback

andrewoverton commented:
Unless I’m forgetting something (which, let’s be real, is entirely possible!) I’ve addressed all the feedback on this, with one exception: I haven’t written tests. I tried doing a unit test, but ran into problems when I realized I couldn’t mock size classes. I then briefly looked into UI tests, but I guess there are some uncertainties with where we stand on those. My understanding is that Earl Grey v1 is not something we want to move forward with, and XCUITest infrastructure should probably be set up in its own PR. So, I’ve added [a comment](https://github.com/material-components/material-components-ios/issues/3823#issuecomment-390784591) in https://github.com/material-components/material-components-ios/issues/3823 to eventually add test coverage to this.

Here are a few more screenshots, because they can't hurt:

![simulator screen shot - ipad air 2 - 2018-05-21 at 17 04 26](https://user-images.githubusercontent.com/8020010/40330354-a6dd7072-5d1a-11e8-8b8d-34e6d0b77637.png)
![simulator screen shot - ipad air 2 - 2018-05-21 at 17 04 29](https://user-images.githubusercontent.com/8020010/40330355-a6ec50d8-5d1a-11e8-98ac-ddf568c9b914.png)
![simulator screen shot - iphone x - 2018-05-21 at 17 01 32](https://user-images.githubusercontent.com/8020010/40330356-a7007b08-5d1a-11e8-9464-5866445f7085.png)
![simulator screen shot - iphone x - 2018-05-21 at 17 01 37](https://user-images.githubusercontent.com/8020010/40330357-a71ad5ca-5d1a-11e8-9f30-c822da286427.png)

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -246,51 +250,41 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+          self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassRegular;
+      BOOL isRegularHeight =
+          self.traitCollection.verticalSizeClass == UIUserInterfaceSizeClassRegular;
+      if (isRegularWidth && isRegularHeight) {
```
If you want to make it easier to unit test, consider refactoring this check into a method that can be (subclassed and) overridden (or partially mocked).

For example, 

```objectivec
- (BOOL)permitsLeadingPresentation {
  // Normally do the trait collection checks here
}

- (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
  ...
  if ([self permitsLeadingPresentation]) {
    ...
  }
}
```

With that, you could hopefully test/verify the behavior of `setSnackbarView:` without requiring actual trait collections to exist.

romoore replied:
The other option is to file an Issue in the backlog and come back to it later.

romoore replied:
Ah, just saw you referenced #3823

romoore commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -32,6 +33,19 @@
+
+ @note The setter must be called from the main thread.
+ */
+@property (class, nonatomic) MDCSnackbarAlignment alignment;
```
Super-nit: Please mark this property as explicitly `assign`.

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -27,26 +29,26 @@
 
-// How far from the bottom of the screen should the snackbar be.
+// How far from the bottom of the screen should the Snackbar be.
 static const CGFloat MDCSnackbarBottomMargin_iPhone = 8.f;
```
Are these constants still based on the device idiom or should they now reflect the size classes?

andrewoverton replied:
I didn't address the bottom margin in this PR--I'll make an issue to make sure that the Snackbar makes no reference to device idioms down the road.

romoore commented on components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h:
```diff
@@ -51,14 +51,26 @@ typedef void (^MDCSnackbarMessageDismissHandler)(BOOL userInitiated,
+ The layout constraint which determines how far the Snackbar is from the leading edge of the screen.
+ It is active when the alignment of the parent overlay view is MDCSnackbarAlignmentLeading.
+ */
+@property(nonatomic, nullable) NSLayoutConstraint *leadingConstraint;
```
Super-nit: Please mark this property and `centerConstraint` as explicitly `strong`.

andrewoverton replied:
I moved the constraints like you suggested doing in https://github.com/material-components/material-components-ios/pull/4163#discussion_r189891840 and I'm noticing that a lot of the strong properties surrounding them in their new location aren't explicitly declared strong. I'm torn between following your advice and marking these two constraints explicitly as strong and following Jeff's advice by keeping consistent with what's there 😬

romoore replied:
I think since it's moving to a class with existing properties, follow the norms there. My previous recommendation was based on my (incorrect?) observation that there seemed to be no previous precedent.

romoore commented on components/Snackbar/src/private/MDCSnackbarOverlayView.m:
```diff
@@ -246,51 +250,41 @@ - (void)setSnackbarView:(MDCSnackbarMessageView *)snackbarView {
+      BOOL isRegularHeight =
+          self.traitCollection.verticalSizeClass == UIUserInterfaceSizeClassRegular;
+      if (isRegularWidth && isRegularHeight) {
+        snackbarView.centerConstraint =
```
Style/nit: It feels a little weird to have an externally-available constraint property on the message view object binding it to the presenting container. Would it work to instead track it in the container?

andrewoverton replied:
Sure, I'll do that

andrewoverton committed:
Move snackbar view constraints to containing view

ianegordon reviewed (approved):
Whoho!

romoore commented:
👏 🎆 👏
